alacritty_terminal/src/tty/windows/winpty.rs

```diff
@@ -31,42 +31,40 @@ use crate::term::SizeInfo;
 use crate::tty::windows::child::ChildExitWatcher;
 use crate::tty::windows::Pty;
 
-// We store a raw pointer because we need mutable access to call
-// on_resize from a separate thread. Winpty internally uses a mutex
-// so this is safe, despite outwards appearance.
-pub struct Agent<'a> {
-    winpty: *mut Winpty<'a>,
+pub struct WinptyAgent {
+    winpty: Arc<Winpty>,
+    pub conout: NamedPipe,
+    pub conin: NamedPipe,
 }
 
-/// Handle can be cloned freely and moved between threads.
-pub type WinptyHandle<'a> = Arc<Agent<'a>>;
-
-// Because Winpty has a mutex, we can do this.
-unsafe impl<'a> Send for Agent<'a> {}
-unsafe impl<'a> Sync for Agent<'a> {}
-
-impl<'a> Agent<'a> {
-    pub fn new(winpty: Winpty<'a>) -> Self {
-        Self { winpty: Box::into_raw(Box::new(winpty)) }
+impl WinptyAgent {
+    pub fn new(winpty: Winpty, conout: NamedPipe, conin: NamedPipe) -> Self {
+        Self { winpty: Arc::new(winpty), conout, conin }
     }
 
-    /// Get immutable access to Winpty.
-    pub fn winpty(&self) -> &Winpty<'a> {
-        unsafe { &*self.winpty }
+    pub fn resize_handle(&self) -> WinptyResizeHandle {
+        WinptyResizeHandle { winpty: self.winpty.clone() }
     }
+}
 
-    pub fn resize(&self, size: &SizeInfo) {
-        // This is safe since Winpty uses a mutex internally.
-        unsafe {
-            (&mut *self.winpty).on_resize(size);
-        }
-    }
+/// Resize handle to safely move between threads.
+pub struct WinptyResizeHandle {
+    // Doesn't need the in/out pipes here!
+    winpty: Arc<Winpty>,
 }
 
-impl<'a> Drop for Agent<'a> {
-    fn drop(&mut self) {
-        unsafe {
-            Box::from_raw(self.winpty);
+impl OnResize for WinptyResizeHandle {
+    fn on_resize(&mut self, size: &SizeInfo) {
+        let (cols, lines) = (size.cols().0, size.lines().0);
+        if cols > 0 && cols <= u16::MAX as usize && lines > 0 && lines <= u16::MAX as usize {
+            let winpty: &mut Winpty = unsafe {
+                // This transmute is actually thread-safe since Winpty uses a mutex internally.
+                std::mem::transmute(&self.winpty as *const _ as *mut Winpty)
+            };
+
+            winpty
+                .set_size(cols as u16, lines as u16)
+                .unwrap_or_else(|_| info!("Unable to set winpty size, did it die?"));
```

Comment:
Why do we need two WinPTY `OnResize` handlers?

Reply:
We don't; this is the one we do need; the old one `impl OnResize for Winpty` should be deleted. I'll tidy up.